File descr alarm v01

Ancairon opened the pull request:
<!--
Describe the change in summary section, including rationale and design decisions.
Include "Fixes #nnn" if you are fixing an existing issue.

In "Component Name" section write which component is changed in this PR. This
will help us review your PR quicker.

In "Test Plan" provide enough detail on how you plan to test this PR so that a reviewer can validate your tests. If our CI covers sufficient tests, then state which tests cover the change.

If you have more information you want to add, write them in "Additional
Information" section. This is usually used to help others understand your
motivation behind this change. A step-by-step reproduction of the problem is
helpful if there is no related issue.
-->

##### Summary
WIP Created the alarm mentioned in #3202 regarding file descriptor limit 

##### Component Name
Collectors and Health 
##### Test Plan

2. I modified the apps_plugin.c to get another dimension on the `netdata.apps_sizes` chart so we can pass it then to an alarm.
3. Made a new alarm file so we can use the dimension mentioned previously.

<!---
Provide enough detail so that your reviewer can understand which test-cases you
have covered, and recreate them if necessary. If sufficient tests are covered
by our CI, then state which tests cover the change.
-->

vlvkobal commented:
`netdata.apps_sizes` is the wrong place to put `fdperc`. Its units are `files/s`, not percentage. You need `fdperc` only for alarm, therefore, you should create and use a health variable, not a dimension. Look at `rrdsetvar_custom_chart_variable_create()` and `nic_speed_max`.

Ancairon commented:
> `netdata.apps_sizes` is the wrong place to put `fdperc`. Its units are `files/s`, not percentage. You need `fdperc` only for alarm, therefore, you should create and use a health variable, not a dimension. Look at `rrdsetvar_custom_chart_variable_create()` and `nic_speed_max`.

Great! Thanks for the comment. If I understand correctly, we need to add `fdperc` as a custom health variable to the `netdata.apps_sizes` chart using `rrdsetvar_custom_chart_variable_create()` and not as a dimension or make a new chart?

thiagoftsm commented on libnetdata/os.c:
```diff
@@ -214,5 +247,4 @@ int getsysctl_by_name(const char *name, void *ptr, size_t len) {
 
-#endif // (TARGET_OS == OS_MACOS)
-
+#endif // (TARGET_OS == OS_MACOS)
```
Github is asking for the last line that you removed here, please, bring it back.

Ancairon replied:
perfect, I added it on commit v3

thiagoftsm commented on health/health.d/apps_plugin.conf:
```diff
@@ -0,0 +1,11 @@
+    crit: $this > (($status == $CRITICAL) ? (85) : (90))
+   delay: down 5m multiplier 1.5 max 1h
+    info: yeet v2
+      to: sysadmin
```
Please, add a new line at the end.

Ancairon replied:
added this too at commit v3

thiagoftsm commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -2191,7 +2193,11 @@ static inline int read_pid_file_descriptors(struct pid_stat *p, void *ptr) {
         }
     }
-
+    usedfdperc = (p->fds_size * 100) / get_system_fd_max();
```
Reading this part of your code, it looks like you need to rename the variables. If something is bigger than the `max`, should I really adjust the max? Before to reach this part of the code, I understood that the maxfdperc, should be a boundary value. :thinking:

Ancairon replied:
I will look into this, I may rename it to `currentmax...` or something similar. `usedfdperc` is the used file descriptors percentage. As mentioned in #3202 we should get the max value of the file descriptors used so `maxfdperc` is the max percentage that I gathered from the PIDs and then I shall pass that into the alarm. Will edit the names to better display the meaning of each variable!

Ancairon commented:
> `netdata.apps_sizes` is the wrong place to put `fdperc`. Its units are `files/s`, not percentage. You need `fdperc` only for an alarm, therefore, you should create and use a health variable, not a dimension. Look at `rrdsetvar_custom_chart_variable_create()` and `nic_speed_max`.

@vlvkobal 
I looked into `rrdsetvar_custom_chart_variable_create()` and `nic_speed_max`, the first can't be implemented without doing `#include "../../database/rrd.h"` but even when I did that and then called the function I had errors while compiling:

/usr/bin/ld: /tmp/apps.plugin.vpXoXi.ltrans0.ltrans.o: in function `main':
<artificial>:(.text.startup+0x1a0f): undefined reference to `rrdsetvar_custom_chart_variable_create'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:3333: apps.plugin] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/home/fotis/netdata'
make[1]: *** [Makefile:5511: all-recursive] Error 1
make[1]: Leaving directory '/home/fotis/netdata'
make: *** [Makefile:2988: all] Error 2

Any idea why this is happening??

Ancairon commented:
Right now the pr is failing the changes due to the issue mentioned above, @thiagoftsm do you have a suggestion on what we should do?

vlvkobal commented:
I'm sorry, somehow it had escaped my attention that you are changing an external plugin. My bad :confused:. You can't use `rrdsetvar_custom_chart_variable_create()`, you should use the external plugins API. Set the variable with [VARIABLE](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#variable).

Ancairon commented:
> I'm sorry, somehow it had escaped my attention that you are changing an external plugin. My bad . You can't use `rrdsetvar_custom_chart_variable_create()`, you should use the external plugins API. Set the variable with [VARIABLE](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#variable).

No problem, will look into this, i got rrdset kinda working with including rrd.h and rrdsetvar.c but I guess the variable solution is more straightforward!

Ancairon commented:
I have made a new commit, fixed variable names to be more undestandable, got the variable in the chart, 
 and got the alarm working! Only thing now is that in both my vm and my custom installation of linux on a laptop the `file-max` file located on `/proc/sys/fs/file-max` is a pretty big number (and possibly cant be parsed corrently on to a  long double) like the number is `9223372036854775807` so I end up with an alarm tha is showing me -0,000000xxx%. Is there anywhere else the max file descriptor limit can be found? I have searched a bit on the internet and went with the file-max way but it is kind of weird of a number. 

An other way I thought we could do this is to check if the number is smaller than 1, like 0.000xxxx an then just display 0% instead of a long floating point number.

thiagoftsm commented:
You really have an unbelievable value for `file-max`, I confess I am shocked. :open_mouth:  Which Linux distribution are you using?
Did you try to use the function `sysconf()` defined at `unistd.h`?  Its documentation says:

```
 OPEN_MAX - _SC_OPEN_MAX
              The maximum number of files that a process can have open at any time.  Must not be less than _POSIX_OPEN_MAX (20).
```

Ancairon commented:
> You really have an unbelievable value for `file-max`, I confess I am shocked. 😮 Which Linux distribution are you using?
> Did you try to use the function `sysconf()` defined at `unistd.h`? Its documentation says:
> 
> ```
>  OPEN_MAX - _SC_OPEN_MAX
>               The maximum number of files that a process can have open at any time.  Must not be less than _POSIX_OPEN_MAX (20).
> ```

I am running Manjaro KDE flavour on a vmware machine, weird I will agree and will have to make another one sooner or later with normal values for testing. I will look into the function you mention and see if it helps!

Ancairon commented:
> You really have an unbelievable value for `file-max`, I confess I am shocked.  Which Linux distribution are you using?
> Did you try to use the function `sysconf()` defined at `unistd.h`? Its documentation says:
> 
> ```
>  OPEN_MAX - _SC_OPEN_MAX
>               The maximum number of files that a process can have open at any time.  Must not be less than _POSIX_OPEN_MAX (20).
> ```

Checked the function, so the line now looks like `long double usedfdpercentage = (long double) ((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX));` to calculate the final percentage that we use on the alarm. 
`usedfdpercentage` is the value that gets passed onto the alarm
`currentmaxfds` is the max percentage of fds used by our pids

Will commit in a bit and then if there ain't any other problem, I will remove the WIP.

Ancairon commented:
@thiagoftsm @vlvkobal I just pushed a commit, it looks finalized to me, from the testing I did the alarm works as intended, shows normal values (close to 32.4% for my system), should I remove the WIP once the checks have passed??

thiagoftsm commented:
> @thiagoftsm @vlvkobal I just pushed a commit, it looks finalized to me, from the testing I did the alarm works as intended, shows normal values (close to 32.4% for my system), should I remove the WIP once the checks have passed??

Hello @Ancairon ,

Yes, please, remove the WIP when the tests end, I will probably review your work today when the WIP is removed.

Best regards!

vlvkobal commented:
Hi @Ancairon, I rechecked the apps plugin code and it looks like `fds_size` is the wrong number. It isn't the number of open files. Look at `aggregate_pid_fds_on_targets()`. I think you need something like that or just place your calculations there right before `aggregate_fd_on_target()`.

You should also move the variable to the `apps.files` chart. `netdata.apps_sizes` is used for internal apps plugin metrics.

Everything else LGTM. Good job 👍

Ancairon commented:
> Hi @Ancairon, I rechecked the apps plugin code and it looks like `fds_size` is the wrong number. It isn't the number of open files. Look at `aggregate_pid_fds_on_targets()`. I think you need something like that or just place your calculations there right before `aggregate_fd_on_target()`.
> 
> You should also move the variable to the `apps.files` chart. `netdata.apps_sizes` is used for internal apps plugin metrics.
> 
> Everything else LGTM. Good job 👍

Perfect, thanks for the review! I will look into this, adjusting the calculations and moving the variable should be easy

ilyam8 commented:
@Ancairon code freeze on this Friday, this is a nice change and if you want to get it into the new release... 😉

Ancairon commented:
> @Ancairon code freeze is planned on this Friday, this is a nice change and if you want to get it into the new release... 😉 

Perfect, better make it quickly then😁

Ancairon commented:
Perfect, made a

> Hi @Ancairon, I rechecked the apps plugin code and it looks like `fds_size` is the wrong number. It isn't the number of open files. Look at `aggregate_pid_fds_on_targets()`. I think you need something like that or just place your calculations there, right before `aggregate_fd_on_target()`.
> 
> You should also move the variable to the `apps.files` chart. `netdata.apps_sizes` is used for internal apps plugin metrics.
> 
> Everything else LGTM. Good job

Made a commit, moved the calculations right before `aggregate_fd_on_target()` on the `aggregate_pid_fds_on_targets()` function. 
Also moved the variable to the `apps.files` chart of the disk family. 

Everything seems to work, provided the checks pass we should be good!

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -2191,7 +2193,7 @@ static inline int read_pid_file_descriptors(struct pid_stat *p, void *ptr) {
         }
     }
-
+    
```
Please remove trailing spaces.

Ancairon replied:
Removed and recommited

Ancairon replied:
Missed an enter my bad.

vlvkobal commented:
Also, look at [LGTM.com alerts](https://lgtm.com/projects/g/netdata/netdata/rev/pr-0d8690df4987a59dfbeeaf9fc631b343486d8372) you got in https://github.com/netdata/netdata/pull/10192#issuecomment-736624750.

Ancairon commented:
> Also, look at [LGTM.com alerts](https://lgtm.com/projects/g/netdata/netdata/rev/pr-0d8690df4987a59dfbeeaf9fc631b343486d8372) you got in [#10192 (comment)](https://github.com/netdata/netdata/pull/10192#issuecomment-736624750).

I saw the warning, trying to see how I can fix it. Maybe the way I passed the variable to the chart isn't proper?

Ancairon commented:
Found the mistake will recommit!

Ancairon commented:
I overlooked one line from `netdata.apps_sizes` and didn't remove it, also missed `, dt` in the apps.files chart and finally I was castin my variable `long double` instead of the default `unsigned long long`

Tested it and now works fine LGTM should be happy now!

Ancairon commented:
The checks seem to be all queued? The LGTM passed but everything else is on queue

Ancairon commented:
@vlvkobal Any idea why this is happening? The pr is stuck for over an hour now on queued checks...

vlvkobal commented:
https://www.githubstatus.com/
GitHub Actions - Degraded

Ancairon commented:
> https://www.githubstatus.com/
> GitHub Actions - Degraded

Now it canceled one of the checks because it was running for 3 hours, what should I do?

vlvkobal commented:
Let's wait and restart jobs tomorrow.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3006,6 +3008,12 @@ static inline void aggregate_pid_fds_on_targets(struct pid_stat *p) {
             continue;
 
+
+        currentfds = p->fds_size;
```
I wrote that you can't use `fds_size`. You should count for open files here and update `currentmaxfds` after the loop has finished.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3238,6 +3249,11 @@ void send_resource_usage_to_netdata(usec_t dt) {
                     , update_every
                     , RATES_DETAIL
             );
+        fprintf(stdout,
+                "CHART apps.files '' 'Apps Open Files' 'open files' disk apps.files stacked 140004 %1$d\n"
+                "VARIABLE fdperc = usedfdpercentage '' absolute 1 1\n"
+                , update_every
+        );
```
There is a [better place](https://github.com/netdata/netdata/blob/master/collectors/apps.plugin/apps_plugin.c#L3800-L3805) to define the variable. Just check for the chart type to avoid multiple definitions.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3304,6 +3320,14 @@ void send_resource_usage_to_netdata(usec_t dt) {
             , (unsigned int)(cmajflt_fix_ratio * 100 * RATES_DETAIL)
             );
+    
+    fprintf(stdout,
+        "BEGIN apps.files %llu\n"
+        "VARIABLE fdperc = %llu\n"
+        "END\n"
+        , dt
+        , usedfdpercentage
+        );
```
A [better place](https://github.com/netdata/netdata/blob/master/collectors/apps.plugin/apps_plugin.c#L3605-L3609) to send the value of the variable.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3194,9 +3202,12 @@ void send_resource_usage_to_netdata(usec_t dt) {
         memmove(&me_last, &me, sizeof(struct rusage));
     }
+    
+    unsigned long long usedfdpercentage = (unsigned long long) ((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX));
```
Do it in `send_collected_data_to_netdata()`.

Ancairon commented:
@vlvkobal I checked the review, I am trying to pass the variable to the chart in the way you mentioned, but can't seem to get the alarm initialized... should I share with you the lines I added to the sectors you mentioned? Maybe I am missing something simple. I also saw the first review, but will change the calculations after I get the alarm working again.

vlvkobal commented:
> should I share with you the lines I added to the sectors you mentioned?

Yes, please.

Ancairon commented:
> > should I share with you the lines I added to the sectors you mentioned?
> 
> Yes, please.

Perfect, thanks for the quick respond! Here is what I have right now:

First of, we need to define the variable, so I did it like this:
```
 if(enable_file_charts) {
           fprintf(stdout, "CHART %s.files '' '%s Open Files' 'open files' disk %s.files stacked 20050 %d\n", type,
                       title, type, update_every);
            fprintf(stdout, "VARIABLE fdperc '' absolute 1 1\n");
        for (w = root; w; w = w->next) {
            if (unlikely(w->exposed))
                fprintf(stdout, "DIMENSION %s '' absolute 1 1\n", w->name);
        }
```

After this we need to pass a value to the variable,
`SEND_SET` wants a `const char *name` and a `kernel_uint_t` so I make my value and variable name accordingly. (I have also tried to do it like before giving the name as `"fdperc"` and the value as `unsigned long long`). 

_code for that:_ 
 
```
kernel_uint_t usedfdpercentage = (kernel_uint_t) ((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX));
const char *fdname = "fdperc";
if(enable_file_charts) {
        send_BEGIN(type, "files", dt);
        send_SET(fdname, usedfdpercentage);
        for (w = root; w; w = w->next) {
            if (unlikely(w->exposed && w->processes))
                send_SET(w->name, w->openfiles);
        }
        send_END();
```
I have tried some other layouts also with moving some lines around but it didn't work...

vlvkobal commented:
I would write something like that
```
        fprintf(stdout, "CHART %s.files '' '%s Open Files' 'open files' disk %s.files stacked 20050 %d\n", type,
                       title, type, update_every);
        for (w = root; w; w = w->next) {
            if (unlikely(w->exposed))
                fprintf(stdout, "DIMENSION %s '' absolute 1 1\n", w->name);
        }
        if (!strcmp("apps", type))
            fprintf(stdout, "VARIABLE fdperc '' absolute 1 1\n");
```

```
        send_BEGIN(type, "files", dt);
        for (w = root; w; w = w->next) {
            if (unlikely(w->exposed && w->processes))
                send_SET(w->name, w->openfiles);
        }
        if (!strcmp("apps", type))
            fprintf(stdout, "VARIABLE fdperc = " KERNEL_UINT_FORMAT "\n", usedfdpercentage);
        send_END();
```

vlvkobal commented:
Or even
```
        if (!strcmp("apps", type)) {
            fprintf(
                stdout, "VARIABLE fdperc = " KERNEL_UINT_FORMAT "\n",
                (kernel_uint_t)((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX)));
        }
```

Ancairon commented:
Perfect, I will check it when I have some free time, later today!

Ancairon commented:
Fantastic, thanks a lot i works like a treat! Now only thing is to check how  I calculate the files, you said
> I wrote that you can't use `fds_size`. You should count for open files here and update `currentmaxfds` after the loop has finished.

I have searched a bit for the correct variable, `p->fds_size` is the size of the fds array, isn't that what we are searching for? What should we use instead?

vlvkobal commented:
I think this is the correct way to do calculations
```
    long double currentfds = 0;
    size_t c, size = p->fds_size;
    struct pid_fd *fds = p->fds;
    for(c = 0; c < size ;c++) {
        int fd = fds[c].fd;

        if(likely(fd <= 0 || fd >= all_files_size))
            continue;

        currentfds++;

        aggregate_fd_on_target(fd, w);
        aggregate_fd_on_target(fd, u);
        aggregate_fd_on_target(fd, g);
    }

    if (currentfds >= currentmaxfds)
        currentmaxfds = currentfds;
```
Don't forget to reset `currentmaxfds` on every global iteration.

Ancairon commented:
Made a commit, I judged that I should reset the `currentmaxfds` in the main function right before the function call. When I put it inside the for loop in `calculate_netdata_statistics` inside the `if(enable_file_charts)` it gave me a static 8% while in the main it had some life between 23 and 22 for my system.

Ancairon commented:
Ok will look into this.

Ancairon commented:
> I think this is the correct way to do calculations
> 
> ```
>     long double currentfds = 0;
>     size_t c, size = p->fds_size;
>     struct pid_fd *fds = p->fds;
>     for(c = 0; c < size ;c++) {
>         int fd = fds[c].fd;
> 
>         if(likely(fd <= 0 || fd >= all_files_size))
>             continue;
> 
>         currentfds++;
> 
>         aggregate_fd_on_target(fd, w);
>         aggregate_fd_on_target(fd, u);
>         aggregate_fd_on_target(fd, g);
>     }
> 
>     if (currentfds >= currentmaxfds)
>         currentmaxfds = currentfds;
> ```
> 
> Don't forget to reset `currentmaxfds` on every global iteration.

The problem was that in the code you did `long double current fds = 0` while this is aready declared, just removed the long double, should work!

vlvkobal commented:
It's better to remove the global declaration. You don't need this variable in the global scope.

vlvkobal commented:
Please remove doubled newlines and trailing spaces.

Ancairon commented:
> Please remove doubled newlines and trailing spaces.

On it.

Ancairon commented:
Missed an enter on one line recommiting...

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3197,6 +3204,7 @@ void send_resource_usage_to_netdata(usec_t dt) {
 
     static char created_charts = 0;
     if(unlikely(!created_charts)) {
+        
```
Delete the newline.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3601,12 +3609,17 @@ static void send_collected_data_to_netdata(struct target *root, const char *type
     }
     send_END();
 
+
```
Delete the newline.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3601,12 +3609,17 @@ static void send_collected_data_to_netdata(struct target *root, const char *type
 
+
+    kernel_uint_t usedfdpercentage = (kernel_uint_t) ((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX));
+    const char *fdname = "fdperc";
```
You don't need this variable anymore. Please delete it.

Ancairon replied:
I think github just bugged out, if you mentioned the comment, I removed it in a previous commit!

Ancairon commented:
I think I catched all the trailing spaces, newlines and tabs where they weren't needed.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -4183,7 +4198,7 @@ int main(int argc, char **argv) {
             printf("DISABLE\n");
             exit(1);
         }
-
```
Why you deleted this newline?

Ancairon replied:
For me the line is showing fine, maybe you saw an oudated commit? (sorry I commeted quickly)

vlvkobal commented:
Well, you are right. GitHub showed me an outdated diff.

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3601,12 +3609,17 @@ static void send_collected_data_to_netdata(struct target *root, const char *type
 
+
+    kernel_uint_t usedfdpercentage = (kernel_uint_t) ((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX));
+    const char *fdname = "fdperc";
```
Hm. I still see the variable.

vlvkobal replied:
Please check.

Ancairon replied:
Will do

vlvkobal commented:
Look at https://github.com/netdata/netdata/pull/10192#discussion_r535217137.

Ancairon commented:
> Look at [#10192 (comment)](https://github.com/netdata/netdata/pull/10192#discussion_r535217137).

If you mean the line where the percentage calculation is happening, dont we need it to present a percentage on the alarm? I am asking because on a previous commit there was also a trailing comment there

vlvkobal commented:
I was talking about `const char *fdname = "fdperc";`

Ancairon commented:
> I was talking about `const char *fdname = "fdperc";`

Perfect, removing it.

Ancairon commented:
I am seeing a weird duplicate somewhere in the code, Im going to look into it and mention the line

Ancairon commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3803,6 +3813,8 @@ static void send_charts_updates_to_netdata(struct target *root, const char *type
                 fprintf(stdout, "DIMENSION %s '' absolute 1 1\n", w->name);
         }
+        if (!strcmp("apps", type))
+            fprintf(stdout, "VARIABLE fdperc '' absolute 1 1\n");
```
Over here I never recall adding this, maybe I should remove it (might mislooked it somehow)

Ancairon commented:
Im sorry, I must have overlooked a part of the code and it had a duplicate somewhere, resolved that. Again sorry for the burst of commits!

vlvkobal commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -3601,12 +3608,15 @@ static void send_collected_data_to_netdata(struct target *root, const char *type
     }
     send_END();
 
+    kernel_uint_t usedfdpercentage = (kernel_uint_t) ((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX));
```
One last thing. You can reduce the scope of this variable. Place it between 3618 and 3619. Or use `(kernel_uint_t) ((currentmaxfds * 100) / sysconf(_SC_OPEN_MAX))` directly inside `fprintf()`.

Ancairon replied:
Moved it inside the if condition!

Ancairon commented:
hmmm one failed lets see why

Ancairon commented:
fixed.

vlvkobal commented:
@Ancairon thank you for your contribution.

Ancairon commented:
> @Ancairon thank you for your contribution.

Thanks a lot for the help, this was quite the pr!

thiagoftsm reviewed (approved):
Thank you @Ancairon !
